pulp_container/app/registry_api.py

```diff
@@ -844,83 +833,128 @@ def put(self, request, path, pk=None):
-        except models.Blob.DoesNotExist:
-            raise BlobInvalid(digest=config_digest)
-
-        # both docker/oci format should contain layers, digest, media_type, size
-        layers = content_data.get("layers")
-        blobs = set()
-        for layer in layers:
-            media_type = layer.get("mediaType")
-            urls = layer.get("urls")
-            digest = layer.get("digest")
-            if (
-                media_type
-                in (
-                    models.MEDIA_TYPE.FOREIGN_BLOB,
-                    models.MEDIA_TYPE.FOREIGN_BLOB_OCI,
+            manifest_list = self._save_manifest(artifact, manifest_digest, request)
+
+            manifests = {}
+            for manifest in content_data.get("manifests"):
+                manifests[manifest["digest"]] = manifest["platform"]
+
+            digests = set(manifests.keys())
+            found_manifests = models.Manifest.objects.filter(digest__in=digests)
+
+            found_digests = set(found_manifests.values_list("digest", flat=True))
+            digests_not_found = digests - found_digests
+            if digests_not_found:
+                ManifestNotFound(reference=digests_not_found)
+
+            manifests_to_list = []
+            for manifest in found_manifests:
+                platform = manifests[manifest.digest]
+                manifest_to_list = models.ManifestListManifest(
+                    manifest_list=manifest,
+                    image_manifest=manifest_list,
+                    architecture=platform["architecture"],
+                    os=platform["os"],
+                    features=platform.get("features", ""),
+                    variant=platform.get("variant", ""),
+                    os_version=platform.get("os.version", ""),
+                    os_features=platform.get("os.features", ""),
                 )
-                and not urls
-            ):
-                raise ManifestInvalid(digest=manifest_digest)
-            if media_type not in (
-                models.MEDIA_TYPE.REGULAR_BLOB,
-                models.MEDIA_TYPE.REGULAR_BLOB_OCI,
+                manifests_to_list.append(manifest_to_list)
+
+            models.ManifestListManifest.objects.bulk_create(
+                manifests_to_list, ignore_conflicts=True, batch_size=1000
+            )
+            manifest = manifest_list
+        else:
+            # both docker/oci format should contain config, digest, mediaType, size
+            config_layer = content_data.get("config")
+            config_media_type = config_layer.get("mediaType")
+            config_digest = config_layer.get("digest")
+            if config_media_type not in (
+                models.MEDIA_TYPE.CONFIG_BLOB,
+                models.MEDIA_TYPE.CONFIG_BLOB_OCI,
             ):
-                raise BlobInvalid(digest=digest)
-            blobs.add(digest)
+                raise BlobInvalid(digest=config_digest)
+
+            try:
```

Comment:
why this is needed? you can just use the repository object directly from line 843